llvm/include/llvm/CAS/CASDB.h

```diff
@@ -177,24 +134,212 @@ class NodeRef : public ObjectRef {
+/// another
+///   ObjectStore and provides an in-memory view of it. Guarantees in-memory
+///   lifetime for objects (via classes \c InMemoryBlob, \c InMemoryTree, and \c
+///   InMemoryNode). Re-exports the low-level ObjectStore APIs and adds new
+///   high-level APIs; uses the "Capture" out-parameters to get memory mapped
+///   regions from the adapted ObjectStore. Takes on the \a Namespace of the
+///   adapted ObjectStore.
+///     - \c InMemoryObject. Base class for objects. Has UniqueID (Namespace /
+///       hash size / hash). Probably change \a ObjectRef to wrap a non-null \c
+///       InMemoryObject*.
+///     - \c InMemoryBlob : InMemoryObject. Adds a StringRef for Data.
+///     - \c InMemoryNode : InMemoryObject. Adds a StringRef for Data. Adds
+///       ArrayRef containing a flat array of hashes (which combines with
+///       Namespace to make a FlatUniqueIDArrayRef).
+///     - \c InMemoryTree : InMemoryObject. Simplest to use a specific format
+///       and translate from underlying CAS object to make the in-memory view.
+///     - get{Blob,Node,Tree}    => InMemoryBlob*,InMemoryNode*,InMemoryTree*
+///       (returns Expected<*>; missing is nullptr)
+///     - expect{Blob,Node,Tree} => InMemoryBlob&,InMemoryNode&,InMemoryTree&
+///       (returns Expected<&>; missing is an error)
+///     - create{Blob,Node,Tree} => InMemoryBlob&,InMemoryNode&,InMemoryTree&
+///       (returns Expected<&>; missing is an error)
+///     - Eventually: overloads with continuation out-param
+///     - Eventually: overloads with std::future out-param
+///     - Examples:
+///          - If an InMemoryView adapts a BuiltinDatabaseCAS, it should be
+///            equivalent to the current return of \a createOnDiskCAS().
+///          - If an InMemoryView adapts a BuiltinNullCAS, it should be
+///            equivalent to the current return of \a createInMemoryCAS().
+///
+/// - \a ActionCache (which has been split out!).
+///
+/// Plugins:
+/// ========
+///
+/// - InMemoryView does not have plugins
+///     - Adapts other (concrete) ObjectStores
+///     - Stores mapped memory regions, as necessary
+///     - Provides views of trees/blobs/etc.
+///
+/// - ObjectStorePlugin : ObjectStore
+///     - Move BuiltinCAS and subclasses into a plugin
+///     - Plugins should be out-of-process (in a daemon)
+///     - Use memory mapped IPC to avoid overhead / keep BuiltinCAS fast
+///     - Maybe: add an "in-memory store" plugin that's ephemeral, lasting just
+///       until the daemon shuts down but sharing work as long as the daemon
+///       lives
+///
+/// - ActionCachePlugin : ActionCache
+///     - Move OnDiskActionCache into a plugin
+///     - Plugins should be out-of-process (in a daemon)
+///     - Maybe: add an "in-memory cache" plugin that's ephemeral, lasting just
+///       until the daemon shuts down but sharing work as long as the daemon
+///       lives
+///
+/// "Capture" out-parameters:
```

Comment:
Yes, it is a bit complicated.

Note that `CASDB` dies and is replaced by `InMemoryView` (which has a high-level API) and `ActionCache` (for caching).

What you're commenting on here are low-level `ObjectStore` APIs (which `InMemoryView` will layer on top of). But you're referring to a `CASDB`. What do you mean by `CASDB`?

Assuming you mean `ObjectStore`: the eventual goal is to move the underlying `ObjectStore` out of process (into a daemon), with object content passed via shared memory. Since the protocol has not been designed, it's a bit hard to reason about. I'm trying to leave room for something fairly low-level that doesn't require per-object allocations and reference counting. For example, perhaps small objects can most efficiently be sent across the process boundary by using a large-ish, sharded, shared memory region that's used as a scratch buffer (or, more simply, over the socket), whereas large objects would have a dedicated mmap'ed memory region.

> for example, don't invalidate the tree while it is being visited

This sounds like a concern for a user of `InMemoryView`; WDYT?

Reply:
Didn't not aware of all the terminology changes. But yes, I mean ObjectStore.

The main concern I have is that if the compiler need to constantly aware of which low level protocol it needs to use for different object vs. the decision is entirely made by the CAS/plugin. I would highly prefer the later since that makes the life of people who do not want to think about CAS easier, have a consistent interface which is easier to wrapper under something like a file system, and much easier to scale with different plugins.

Reply:
I expect there to be almost no direct users of `ObjectStore`: instead, it'll be adapted by an `InMemoryView` to provide lifetime. Unit tests may use the object store directly for testing the low-level interface, and maybe there will be other special cases.

It sounds to me like the use cases you're describing would be interacting with `InMemoryView`. WDYT?

Reply:
That means InMemoryView need to be tuned for different plugins? That sounds better but I am not sure if those decisions should be made within LLVM.

Reply:
The client (usually `InMemoryView`) doesn't have to tune. The design of the low-level `ObjectStore` APIs is that the client passes a `Capture` object  (usually `InMemoryView`) which includes the capabilities the client has (for blobs, this could include `raw_ostream&` (copy bytes), `StringRef &` (permanent lifetime), `mapped_file_region` (ownership), etc.) and an indication of its preferences (e.g., only use `mapped_file_region` for blobs between 16K and 2GB) and then the object store implementation/plugin sends the object over using one of the mechanisms.

For something a bit more concrete, see the incomplete sketch of `BlobCapture`/`TreeCapture`/`NodeCapture` deleted at https://github.com/dexonsmith/llvm-project/commit/6dffd648a4bbd0395f22c871137b5df32a587356#diff-5d292c70f2336e2afef84ce13f5df8b6ad948290677fa37618de819a6c9d3d70, coming from:
  https://github.com/dexonsmith/llvm-project/commits/experimental/cas/refactor-casdb-before-deleting-history/llvm/include/llvm/CAS/CASDB.h